lib/stitches/running_stitch.py

```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
     return [line.interpolate(x, normalized=False) for x in splits]
 
 
-def running_stitch(points, stitch_length, tolerance):
-    """Generate running stitch along a path.
-
-    Given a path and a stitch length, walk along the path in increments of the
-    stitch length.  If sharp corners are encountered, an extra stitch will be
-    added at the corner to avoid rounding the corner.  The starting and ending
-    point are always stitched.
-
-    The path is described by a set of line segments, each connected to the next.
-    The line segments are described by a sequence of points.
-    """
-
+class AngleInterval():
+    # Modular interval containing either the entire circle or less than half of it
+    # partially based on https://fgiesen.wordpress.com/2015/09/24/intervals-in-modular-arithmetic/
+
+    def __init__(self, a: float, b: float, all: bool = False):
+        self.all = all
+        self.a = a
+        self.b = b
+
+    @staticmethod
+    def all():
+        return AngleInterval(0, math.tau, True)
+
+    @staticmethod
+    def fromBall(p: Point, epsilon: float):
+        d = p.length()
+        if d <= epsilon:
+            return AngleInterval.all()
+        center = p.angle()
+        delta = math.asin(epsilon / d)
+        return AngleInterval(center - delta, center + delta)
+
+    @staticmethod
+    def fromSegment(a: Point, b: Point):
+        angleA = a.angle()
+        angleB = b.angle()
+        diff = (angleB - angleA) % tau
+        if diff == 0 or diff == math.pi:
+            return None
+        elif diff < math.pi:
+            return AngleInterval(angleA - 1e-6, angleB + 1e-6)
+        else:
+            return AngleInterval(angleB - 1e-6, angleA + 1e-6)
+
+    def containsAngle(self, angle: float):
+        if self.all:
+            return True
+        return (angle - self.a) % tau <= (self.b - self.a) % tau
+
+    def containsPoint(self, p: Point):
+        return self.containsAngle(math.atan2(p.y, p.x))
+
+    def intersect(self, other):
+        # assume that each interval contains less than half the circle (or all of it)
+        if other is None:
+            return None
+        elif self.all:
+            return other
+        elif other.all:
+            return self
+        elif self.containsAngle(other.a):
```

Comment:
Similar here: the paper's scheme allows this to be simply`AngleInterval(max(self.a, other.a), min(self.b, other.b))`.

Reply:
see response for the other spot. I wanted not to have restrictions on a and b.

Reply:
@george-steel I don't think the paper restricts a and b, at least not in practical terms; it just requires that we add 360 degrees to b if b would be less than a.

Reply:
I did it the other way at first, but the number of cases was becoming really messy and hard to keep track of. Modular comparison makes it a lot tidier.